libtenzir/builtins/operators/export.cpp

```diff
@@ -37,45 +35,36 @@
 namespace tenzir::plugins::export_ {
 
 struct bridge_state {
-  std::queue<table_slice> buffer;
-  caf::typed_response_promise<table_slice> rp;
+  std::queue<table_slice> buffer = {};
+  caf::typed_response_promise<table_slice> rp = {};
+  expression expr = {};
 };
 
-caf::behavior
-make_bridge(caf::stateful_actor<bridge_state>* self, importer_actor importer) {
+caf::behavior make_bridge(caf::stateful_actor<bridge_state>* self,
+                          importer_actor importer, expression expr) {
+  self->state.expr = std::move(expr);
   self
-    ->request(importer, caf::infinite,
-              caf::actor_cast<stream_sink_actor<table_slice>>(self))
-    .then([](caf::outbound_stream_slot<table_slice>) {},
-          [self](caf::error err) {
-            self->quit(err);
+    ->request(importer, caf::infinite, atom::subscribe_v,
+              caf::actor_cast<receiver_actor<table_slice>>(self))
+    .then([]() {},
+          [self](const caf::error& err) {
+            self->quit(add_context(err, "failed to subscribe to importer"));
           });
-
   return {
-    [self](
-      caf::stream<table_slice> in) -> caf::inbound_stream_slot<table_slice> {
-      return caf::attach_stream_sink(
-               self, in,
-               [](caf::unit_t&) {
-                 // nop
-               },
-               [=](caf::unit_t&, table_slice slice) {
-                 if (self->state.rp.pending())
-                   self->state.rp.deliver(std::move(slice));
-                 else
-                   self->state.buffer.push(std::move(slice));
-               },
-               [=](caf::unit_t&, const caf::error& err) {
-                 if (err)
-                   TENZIR_ERROR("{} got error during streaming: {}", *self,
-                                err);
-               })
-        .inbound_slot();
+    [self](table_slice slice) {
+      auto filtered = filter(std::move(slice), self->state.expr);
+      if (not filtered)
+        return;
+      if (self->state.rp.pending()) {
+        self->state.rp.deliver(std::move(*filtered));
+      } else {
+        self->state.buffer.push(std::move(*filtered));
```

Comment:
I think we need to limit the buffer size here. Otherwise we're going to get memory issues when the downstream operator isn't pulling from the `export --live`.

Reply:
In principle it depends on how much is available. If we had topics with bounded buffers it would already be solvable dynamically at the pipeline definition level, but for now I'd say even 100K events should be enough, and be less of a burden for small deployments.

Reply:
We agreed in a call to use a higher buffer size. I chose 4Mi for now, which is also our partition size.